src/pubsub/ua_pubsub.c

```diff
@@ -182,6 +185,1098 @@ UA_Server_removeWriterGroup(UA_Server *server, const UA_NodeId writerGroup){
+                                UA_PubSubReaderGroupConfig *dst) {
+    UA_String_copy(&src->name, &dst->name);
+    //Currently simple memcpy only
+    memcpy(&dst->securityParameters, &src->securityParameters, sizeof(UA_PubSubSecurityParameters));
+    return UA_STATUSCODE_GOOD;
+}
+
+/**
+ * Process NetworkMessage.
+ *
+ * @param server
+ * @param networkmessage
+ * @return UA_STATUSCODE_GOOD on success
+ */
+UA_StatusCode
+UA_PubSubConnection_processNetworkMessage(UA_Server *server, UA_NetworkMessage *pMsg,
+                                          UA_PubSubConnection *pConnection) {
+    UA_StatusCode retval = UA_STATUSCODE_BADNOTIMPLEMENTED;
+    if ((pMsg == NULL) || (pConnection == NULL)) {
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+    }
+
+    UA_Boolean publisherIdAvailabe = false;
+    if (pMsg->publisherIdEnabled) {
+        publisherIdAvailabe = true;
+    }
+
+    UA_Boolean writerGroupIdAvailabe = false;
+    if (pMsg->groupHeaderEnabled && pMsg->groupHeader.writerGroupIdEnabled) {
+        writerGroupIdAvailabe = true;
+    }
+
+    UA_Boolean dataSetWriterIdAvailabe = false;
+    if (pMsg->payloadHeaderEnabled) {
+        dataSetWriterIdAvailabe = true;
+    }
+
+    // DataSetClassId is not necessarily able to identify the corresponding DataSetReader:
+    // all DataSetMessages in the msg have the same DataSetClassId. Anyway it should be checked
+    UA_Boolean dataSetClassIdAvailable = false;
+    if (pMsg->dataSetClassIdEnabled) {
+        dataSetClassIdAvailable = true;
+    }
+
+    // ToDo: Here some filtering is possible.
+    UA_Byte anzDataSets = 1;
+    if (pMsg->payloadHeaderEnabled) {
+        anzDataSets = pMsg->payloadHeader.dataSetPayloadHeader.count;
+    }
+
+    for (UA_Byte iterator = 0; iterator < anzDataSets; iterator++) {
+        UA_PubSubDataSetReader* dataSetReaderErg = NULL;
+        if (publisherIdAvailabe || writerGroupIdAvailabe || dataSetWriterIdAvailabe) {
+            UA_PubSubReaderGroup* readerGroup = NULL;
+            LIST_FOREACH(readerGroup, &pConnection->readerGroups, listEntry) {
+                UA_PubSubDataSetReader *tmpReader;
+                LIST_FOREACH(tmpReader, &readerGroup->readers, listEntry) {
+                    if (publisherIdAvailabe && (tmpReader->config.publisherId.type != NULL)) {
+                        switch (pMsg->publisherIdType) {
+                            case UA_PUBLISHERDATATYPE_BYTE:
+                                if (tmpReader->config.publisherId.type == &UA_TYPES[UA_TYPES_BYTE]) {
+                                    if (pMsg->publisherId.publisherIdByte == (*(UA_Byte*)tmpReader->config.publisherId.data)) {
+                                        UA_LOG_INFO(&server->config.logger, UA_LOGCATEGORY_SERVER, "DataSetReader found with PublisherId");
+                                        dataSetReaderErg = tmpReader;
+                                    }
+                                }
+                                break;
```

Comment:
If you find a dataSetReaderErg, you should not only break the switch-case, but also the LIST_FOREACH loops.
This is not done imho.

If the search is factored out into a dedicated method, you can just return after the first match.